src/nbl/asset/utils/CGLSLCompiler.cpp

```diff
@@ -13,11 +13,126 @@ using namespace nbl;
+                //  or path relative to executable's working directory (""-type).
+                relDir = std::filesystem::path(_requesting_source).parent_path();
+            }
+            std::filesystem::path name = (_type == shaderc_include_type_relative) ? (relDir / _requested_source) : (_requested_source);
+
+            if (std::filesystem::exists(name) && !reqBuiltin)
+                name = std::filesystem::absolute(name);
+
+            if (_type == shaderc_include_type_relative)
+                res_str = m_defaultIncludeFinder->getIncludeRelative(relDir, _requested_source);
+            else //shaderc_include_type_standard
+                res_str = m_defaultIncludeFinder->getIncludeStandard(relDir, _requested_source);
+
+            if (!res_str.size()) {
+                const char* error_str = "Could not open file";
+                res->content_length = strlen(error_str);
+                res->content = new char[res->content_length + 1u];
+                strcpy(const_cast<char*>(res->content), error_str);
+                res->source_name_length = 0u;
+                res->source_name = "";
+            }
+            else {
+                //employ encloseWithinExtraInclGuards() in order to prevent infinite loop of (not necesarilly direct) self-inclusions while other # directives (incl guards among them) are disabled
+                IShaderCompiler::disableAllDirectivesExceptIncludes(res_str);
+                CGLSLCompiler::disableGlDirectives(res_str);
+                res_str = IShaderCompiler::encloseWithinExtraInclGuards(std::move(res_str), m_maxInclCnt, name.string().c_str());
+
+                res->content_length = res_str.size();
+                res->content = new char[res_str.size() + 1u];
+                strcpy(const_cast<char*>(res->content), res_str.c_str());
+                res->source_name_length = name.native().size();
+                res->source_name = new char[name.native().size() + 1u];
+                strcpy(const_cast<char*>(res->source_name), name.string().c_str());
+            }
+
+            return res;
+        }
+
+        void ReleaseInclude(shaderc_include_result* data) override
+        {
+            if (data->content_length > 0u)
+                delete[] data->content;
+            if (data->source_name_length > 0u)
+                delete[] data->source_name;
+            delete data;
+        }
+    };
+}
+
 CGLSLCompiler::CGLSLCompiler(core::smart_refctd_ptr<system::ISystem>&& system)
     : IShaderCompiler(std::move(system))
 {
 }
 
+
+
+std::string CGLSLCompiler::preprocessShader(std::string&& code, IShader::E_SHADER_STAGE& stage, const SPreprocessorOptions& preprocessOptions) const
+{
+    if (preprocessOptions.extraDefines.size())
+    {
+        insertExtraDefines(code, preprocessOptions.extraDefines);
+    }
+    if (preprocessOptions.includeFinder != nullptr)
```

Comment:
UNRESOLVED: https://github.com/Devsh-Graphics-Programming/Nabla/pull/437/files#r1089687024

If you recall we had a similar logic issue with HLSL compiler I asked you to fix but I see it hasn't been resolved yet.

We should be able to preprocess even if the includeFinder is nullptr.
- pragma shader stage should still work with no includeFinder
- preprocessing with enabling and renabling directives should still be done with no includeFinder

you can just remove this if statement here and make it more specific when you try to actually include something or set the includer.

If an include is requested with `includeFinder==nullptr`, return an empty string and log an error.
If you don't have access to a logger just assert with a message and put a TODO to log a warning/error later